[Flaky unit test] TestEndpointResolver

I just observed the test failing in CI, after a push to `main`.
Since I introduced this test, I will try to fix it :)

```
I0618 20:04:49.808687   25686 endpoint_resolver.go:194] "Starting controller" name="ServiceEndpointResolver:kube-system/antrea"
I0618 20:04:49.808848   25686 shared_informer.go:311] Waiting for caches to sync for ServiceEndpointResolver:kube-system/antrea
I0618 20:04:49.909926   25686 shared_informer.go:318] Caches are synced for ServiceEndpointResolver:kube-system/antrea
E0618 20:04:49.910524   25686 endpoint_resolver.go:245] "Cannot resolve endpoint because of missing resource" err="service \"antrea\" not found" service="kube-system/antrea"
I0618 20:04:51.810642   25686 endpoint_resolver.go:208] "Shutting down controller" name="ServiceEndpointResolver:kube-system/antrea"
I0618 20:04:51.810963   25686 endpoint_resolver.go:194] "Starting controller" name="ServiceEndpointResolver:kube-system/antrea"
I0618 20:04:51.811033   25686 shared_informer.go:311] Waiting for caches to sync for ServiceEndpointResolver:kube-system/antrea
I0618 20:04:51.911733   [256](https://github.com/antrea-io/antrea/actions/runs/9571408711/job/26388477866#step:4:257)86 shared_informer.go:318] Caches are synced for ServiceEndpointResolver:kube-system/antrea
I0618 20:04:51.911928   25686 endpoint_resolver.go:276] "Selected a new Endpoint for Service, notifying listeners" service="kube-system/antrea" url="https://172.18.0.3:10349"
I0618 20:04:51.962225   25686 endpoint_resolver.go:276] "Selected a new Endpoint for Service, notifying listeners" service="kube-system/antrea" url="https://172.18.0.4:10349"
I0618 20:04:52.013356   25686 endpoint_resolver.go:208] "Shutting down controller" name="ServiceEndpointResolver:kube-system/antrea"
I0618 20:04:52.013629   25686 endpoint_resolver.go:194] "Starting controller" name="ServiceEndpointResolver:kube-system/antrea"
I0618 20:04:52.013694   25686 shared_informer.go:311] Waiting for caches to sync for ServiceEndpointResolver:kube-system/antrea
I0618 20:04:52.114582   25686 shared_informer.go:318] Caches are synced for ServiceEndpointResolver:kube-system/antrea
I0618 20:04:52.114736   25686 endpoint_resolver.go:276] "Selected a new Endpoint for Service, notifying listeners" service="kube-system/antrea" url="https://172.18.0.3:10349/"
E0618 20:04:52.163946   25686 endpoint_resolver.go:240] "Cannot resolve endpoint because Service is unavailable" err="no endpoints available for service \"antrea\"" service="kube-system/antrea"
I0618 20:04:52.163991   25686 endpoint_resolver.go:278] "Selected no Endpoint for Service, notifying listeners" service="kube-system/antrea"
I0618 20:04:52.214114   25686 endpoint_resolver.go:208] "Shutting down controller" name="ServiceEndpointResolver:kube-system/antrea"
I0618 20:04:52.214308   25686 endpoint_resolver.go:194] "Starting controller" name="ServiceEndpointResolver:kube-system/antrea"
I0618 20:04:52.214374   25686 shared_informer.go:311] Waiting for caches to sync for ServiceEndpointResolver:kube-system/antrea
I0618 20:04:52.314536   25686 shared_informer.go:318] Caches are synced for ServiceEndpointResolver:kube-system/antrea
I0618 20:04:52.314699   25686 endpoint_resolver.go:276] "Selected a new Endpoint for Service, notifying listeners" service="kube-system/antrea" url="https://172.18.0.3:10349/"
E0618 20:04:52.364906   25686 endpoint_resolver.go:245] "Cannot resolve endpoint because of missing resource" err="service \"antrea\" not found" service="kube-system/antrea"
I0618 20:04:52.364959   25686 endpoint_resolver.go:278] "Selected no Endpoint for Service, notifying listeners" service="kube-system/antrea"
--- FAIL: TestEndpointResolver (2.61s)
    --- FAIL: TestEndpointResolver/add_Service_and_Endpoints (2.00s)
        endpoint_resolver_test.go:111: 
            	Error Trace:	/home/runner/work/antrea/antrea/pkg/agent/client/endpoint_resolver_test.go:112
            	            				/opt/hostedtoolcache/go/1.21.0/x64/src/runtime/asm_amd64.s:1650
            	Error:      	Not equal: 
            	            	expected: &url.URL{Scheme:"https", Opaque:"", User:(*url.Userinfo)(nil), Host:"172.18.0.3:10349", Path:"", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}
            	            	actual  : (*url.URL)(nil)
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,14 +1,2 @@
            	            	-(*url.URL)({
            	            	- Scheme: (string) (len=5) "https",
            	            	- Opaque: (string) "",
            	            	- User: (*url.Userinfo)(<nil>),
            	            	- Host: (string) (len=16) "172.18.0.3:10349",
            	            	- Path: (string) "",
            	            	- RawPath: (string) "",
            	            	- OmitHost: (bool) false,
            	            	- ForceQuery: (bool) false,
            	            	- RawQuery: (string) "",
            	            	- Fragment: (string) "",
            	            	- RawFragment: (string) ""
            	            	-})
            	            	+(*url.URL)(<nil>)
            	            	 
        endpoint_resolver_test.go:111: 
            	Error Trace:	/home/runner/work/antrea/antrea/pkg/agent/client/endpoint_resolver_test.go:111
            	Error:      	Condition never satisfied
            	Test:       	TestEndpointResolver/add_Service_and_Endpoints
I0618 20:04:52.414997   25686 endpoint_resolver.go:208] "Shutting down controller" name="ServiceEndpointResolver:kube-system/antrea"
FAIL
````

--- a/pkg/agent/client/endpoint_resolver_test.go
+++ b/pkg/agent/client/endpoint_resolver_test.go
@@ -29,6 +29,7 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/cache"
 )
 
 const (
@@ -92,10 +93,17 @@ func getEndpointURL(ip string) *url.URL {
 	}
 }
 
-func runTestEndpointResolver(ctx context.Context, objects ...runtime.Object) (*fake.Clientset, *EndpointResolver) {
+func runTestEndpointResolver(ctx context.Context, t *testing.T, objects ...runtime.Object) (*fake.Clientset, *EndpointResolver) {
 	k8sClient := fake.NewSimpleClientset(objects...)
 	resolver := NewEndpointResolver(k8sClient, testNamespace, testServiceName, testServicePort)
 	go resolver.Run(ctx)
+	// Wait for informers to sync to avoid race condition between List and Watch with the fake clientset.
+	// Note that we cannot call resolver.informerFactory.WaitForCacheSync instead, as that only
+	// waits until *started* informers' caches have been synced, and at this point
+	// resolver.informerFactory.Start may not have been called yet.
+	// We also check the return value of cache.WaitForCacheSync even though it should only be
+	// true if the context was cancelled. which should not happen in our test cases.
+	require.True(t, cache.WaitForCacheSync(ctx.Done(), resolver.serviceListerSynced, resolver.endpointsListerSynced))
 	return k8sClient, resolver
 }
 
@@ -103,7 +111,7 @@ func TestEndpointResolver(t *testing.T) {
 	t.Run("add Service and Endpoints", func(t *testing.T) {
 		ctx, cancelFn := context.WithCancel(context.Background())
 		defer cancelFn()
-		k8sClient, resolver := runTestEndpointResolver(ctx)
+		k8sClient, resolver := runTestEndpointResolver(ctx, t)
 		require.Nil(t, resolver.CurrentEndpointURL())
 		svc, endpoints := getTestObjects()
 		k8sClient.CoreV1().Services(testNamespace).Create(ctx, svc, metav1.CreateOptions{})
@@ -117,7 +125,7 @@ func TestEndpointResolver(t *testing.T) {
 		ctx, cancelFn := context.WithCancel(context.Background())
 		defer cancelFn()
 		svc, endpoints := getTestObjects()
-		k8sClient, resolver := runTestEndpointResolver(ctx, svc, endpoints)
+		k8sClient, resolver := runTestEndpointResolver(ctx, t, svc, endpoints)
 		assert.EventuallyWithT(t, func(t *assert.CollectT) {
 			assert.Equal(t, getEndpointURL(testEndpointIP1), resolver.CurrentEndpointURL())
 		}, 2*time.Second, 50*time.Millisecond)
@@ -132,7 +140,7 @@ func TestEndpointResolver(t *testing.T) {
 		ctx, cancelFn := context.WithCancel(context.Background())
 		defer cancelFn()
 		svc, endpoints := getTestObjects()
-		k8sClient, resolver := runTestEndpointResolver(ctx, svc, endpoints)
+		k8sClient, resolver := runTestEndpointResolver(ctx, t, svc, endpoints)
 		assert.EventuallyWithT(t, func(t *assert.CollectT) {
 			assert.Equal(t, getEndpointURL(testEndpointIP1), resolver.CurrentEndpointURL())
 		}, 2*time.Second, 50*time.Millisecond)
@@ -147,7 +155,7 @@ func TestEndpointResolver(t *testing.T) {
 		ctx, cancelFn := context.WithCancel(context.Background())
 		defer cancelFn()
 		svc, endpoints := getTestObjects()
-		k8sClient, resolver := runTestEndpointResolver(ctx, svc, endpoints)
+		k8sClient, resolver := runTestEndpointResolver(ctx, t, svc, endpoints)
 		assert.EventuallyWithT(t, func(t *assert.CollectT) {
 			assert.Equal(t, getEndpointURL(testEndpointIP1), resolver.CurrentEndpointURL())
 		}, 2*time.Second, 50*time.Millisecond)

